### ISSUE TITLE
Issue with a type of Var when referencing types

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#1591]
`Var` can point to a:

* `Variable`, the `type` field of can be any of (and the `Var` expr node is assumed to have this type): 
  * any primitive type (the `type_declaration` field is null), like an:
    * `Integer`
    * `Real`
    * `Complex`
    * `Logical`
  * any non-primitive type (the `type_declaration` field points to the respective symbol, given below), like a:
    * function pointer (`type=FunctionType`, `type_declaration=Function`)
    * structure instance (`type=StructType`, `type_declaration=Struct`)
    * class instance (`type=ClassType`, `type_declaration=Class`)
    * enumeration instance (`type=EnumType`, `type_declaration=Enum`)
* `Function`: for using some function declaration name in an expression as a value, such as assigning it to a function pointer (`f => my_fn`), or passing it as a callback to a function argument (`call f(my_fn)`). The type of the `Function` is `FunctionType` and that is the type of the `Var` that references it as well.

An issue is how to represent `sizeof(i32)` and `sizeof(struct A)`:

https://github.com/lcompilers/lpython/pull/1730/files#r1173780425

In ASR we have a `SizeOfType(ttype arg, ttype type, expr? value)` expr node, and the argument is a type, so that is clean and there is no problem. The `sizeof()` is a special function, as it does not accept an expression (`expr`), but rather a type (`ttype`).

However, in the link above there is code like:
```c++
ASR::Var_t* arg_Var = ASR::down_cast<ASR::Var_t>(args[0].m_value);
ASR::symbol_t* arg_Var_m_v = ASRUtils::symbol_get_past_external(arg_Var->m_v);
if( ASR::is_a<ASR::StructType_t>(*arg_Var_m_v) ) {
    ...
}
```
Which should never happen, as `Var` can't point to a `Struct` symbol node (the "StructType" in master means the `Struct` symbol node above). So either the code above has to be fixed, or our design has to be fixed.

The fundamental issue here is if we should allow referencing types like Struct or Class as expressions. Our current design only allows referencing their instances ("values"), but not the types themselves. `Function` is allowed, as it behaves like value, of the type `FunctionType`.

**Comment by czgdp1807** [comment]
`StructType` is also a `symbol` in LFortran's `main` branch,

https://github.com/lfortran/lfortran/blob/f82602022df8768b912014fd0c78147cc849823c/src/libasr/ASR.asdl#L82-L99

`Var::m_v` is of `ASR::symbol_t*` type.

**Comment by rebcabin** [comment]
I think it's probably a good "open-door" for types like Struct or Class to be first-class values. Such makes ASR's type system programmable. The Little Typer DSL in Racket does such.

**Comment by rebcabin** [comment]
BTW, I need clarification on `Struct` versus `StructType` etc. Can you post a few memorable words that make the distinction crystal clear, with examples?

**Comment by certik** [comment]
Yes, the door is open to Struct/Class being first class values, but that's for later: every expression has a type, so what type is "Struct A"? Well, it's `ttype`, except that `ttype` is not a type in our current ASR. We need a "type of a type", a meta-type (I think it's similar to metaclasses in Python). I am sure it can be done, but I would recommend to first finish our current design without meta-types, get it done right and clean. And then tackle meta-types.

In the current design without meta-types. In master, the Struct and StrucType names are exactly backwards, so from now on, I am referring to the new design that we will move to, described in the issue description above. The StructType is a ttype, so any expression can have this type, say if you return a struct from a function, then the FunctionCall with have a type = StructType. If you declare a struct variable, then the Var that references the variable will have a type StructType. StructType does not have any names (neither members nor struct name), it only contains the types of the members. StrucType *might* need to contain information about alignment, so that SizeOf above can correctly return the size of the struct just from the ttype. It seems to me any ttype should contain enough information so that the backend can infer exactly the layout in memory just from the ttype (and compiler command line options), it shouldn't need access to Struct, see below.

Struct, on the other hand, is a symbol, that represents the declaration of the struct type in the code. It's scoped, it contains the names of members and the name of the struct. It is referenced from Variable in the type_declaration field. It is used by the frontend to enforce C style struct equality, if needed, as you need to consult either the name, or where it is declared. It is also used by the frontend to figure out which member to assign to, since members are referenced by name in the source code. Struct is thus inherently tied to the symbol table, it is scoped. 

StructType is a ttype, independent of scope, global.

**Comment by certik** [comment]
Another idea, to be documented: ttypes are global, can't reference a symbol table. However, ttypes must allow expressions, to represent sizes of arrays. Essentially dependent types. However, those expressions cannot reference a symbol table. To achieve it, we are doing FunctionParam, to allow referencing a function parameter (argument), without referencing the symbol table. So FunctionParam is global, it just says `(FunctionParam 2)`, meaning the second function argument. I think the only way to reference a symbol table from and expression should be `Var`. And so using `Var` in an expression in ttype is not allowed. But that's it. So I think that's a clean design.

**Comment by certik** [comment]
The only not completely clean part is that Var can points to Function directly. Let's keep it like that for now, but I think later we might create a FunctionValue expr node, that would point to Function directly, allowing the function name to be used in expressions like passing it as a callback (`call f(my_fn)`), and `Var` would only strictly point to a `Variable`.

**Comment by certik** [comment]
Here is another issue: the `SizeOfType(ttype arg, ttype type, expr? value)` expr node is used in the C backend to generate `sizeof(struct A)`. Currently in master, the `ttype` links to the symbol table, so one is able to determine the name "A". With our new design, we would not know the name. So we might need to change the node to `SizeOfType(ttype arg, symbol? type_declaration, ttype type, expr? value)` and put the `Struct` in there. Alternatively, we might need to put the name of the struct as "identifier" into the StructType ttype.

**Comment by czgdp1807** [comment]
Correct. We discussed in our meeting. `SizeOfType` looses some information from AST. So we may need to add an optional node, I guess `identifier` would be a better choice then an optional `expr`. Just store the name of the type, no need to reference to the original symbol.
